File: 0682_Baseball_Game/solution.java
package cn.huang.leetcode;

import java.util.Stack;

/*
    682. Baseball Game

 */
public class LC_0682_BaseballGame {
    public int calPoints(String[] ops) {
        if (ops == null || ops.length == 0)
            return 0;

        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < ops.length; i++) {
            String op = ops[i];
            if (op.equals("C")) {
                stack.pop();
            } else if (op.equals("+")) {
                int score1 = stack.pop();
                int sum = score1 + stack.peek();
                stack.push(score1);
                stack.push(sum);
            } else if (op.equals("D")) {
                stack.push(2 * stack.peek());
            } else {
                stack.push(Integer.parseInt(op));
            }
        }
        int sum = 0;

        while (!stack.empty()) {
            sum += stack.pop();
        }
        return sum;
    }

    public int calPoints2(String[] ops) {
        Stack<Integer> history = new Stack<>();
        for (String op : ops) {
            switch (op) {
                case "+":
                    history.push(history.peek() + history.get(history.size() - 2));
                    break;
                case "D":
                    history.push(history.peek() * 2);
                    break;
                case "C":
                    history.pop();
                    break;
                default:
                    history.push(Integer.parseInt(op));
            }
        }
        int result = 0;
        for (int item : history) {
            result += item;
        }
        return result;
    }
}

File: 0682_Baseball_Game/solution.cpp
/*
682. Baseball Game

Level: Easy

https://leetcode.com/problems/baseball-game
*/
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>

#include <algorithm>
#include <stack>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

class Solution
{
public:
  int calPoints(vector<string> &ops)
  {
    stack<int> stack;

    for (int i = 0; i < ops.size(); i++)
    {
      string op = ops[i];
      if (op == "C")
      {
        stack.pop();
      }
      else if (op == "+")
      {
        int score1 = stack.top();
        stack.pop();
        // sum of the last two valid round's points
        int sum = score1 + stack.top();
        stack.push(score1);
        stack.push(sum);
      }
      else if (op == "D")
      {
        // score is the doubled data of the last valid round's points.
        stack.push(2 * stack.top());
      }
      else
      {
        stack.push(stoi(op));
      }
    }
    int sum = 0;

    while (!stack.empty())
    {
      sum += stack.top();
      stack.pop();
    }
    return sum;
  }
};